### PR TITLE
Add XDC parser to FPGA interchange

### PR DESCRIPTION
**Pull request by litghost**
Builds on #579

### CODE REVIEW

**Comment by litghost** [comment]
This PR should update nextpnr-tests submodule once https://github.com/YosysHQ/nextpnr-tests/pull/3 is merged.

File: fpga_interchange/xdc.cc [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+ */
+
+#include "xdc.h"
+#include "log.h"
+#include "nextpnr.h"
+#include <string>
+
+NEXTPNR_NAMESPACE_BEGIN
+
+static int port_set_from_any(Tcl_Interp *interp, Tcl_Obj *objPtr) {
```
Comment (gatecat) [inline fpga_interchange/xdc.cc:28]: looking at the braces I think a `clangformat` is needed here?

Comment (litghost) [reply fpga_interchange/xdc.cc]: Fixed.

**Comment by litghost** [comment]
> This PR should update nextpnr-tests submodule once [YosysHQ/nextpnr-tests#3](https://github.com/YosysHQ/nextpnr-tests/pull/3) is merged.

I've update the submodule, so the unit test should now be included.

**Comment by litghost** [comment]
I believe this is pretty much ready.  I tried adding the FPGA interchange tests to the CI, but it failed with https://cirrus-ci.com/build/5400920836014080 `Error while loading tasks: FAILED_PRECONDITION: Unknown field test_fpga_interchange on 'build-test-ubuntu2004' YAML node!!`

Commit that didn't work: https://github.com/YosysHQ/nextpnr/commit/f35842c0701da3e8b9c212d3b5c410403e8c3b42

**Comment by gatecat** [comment]
can you try using `test_fpga_interchange_script` as the yaml node, instead of just `test_fpga_interchange`? I'm not that familiar with the details of cirrus but this is the only obvious thing that it could be.

**Comment by gatecat** [comment]
Looks like it picked that up! Let's just check that it passes...

**Comment by litghost** [comment]
> Looks like it picked that up! Let's just check that it passes...

Green, and it ran:
```
cd build && ./nextpnr-fpga_interchange-test
[==========] Running 5 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 2 tests from XdcTest
[ RUN      ] XdcTest.do_nothing_interp
[       OK ] XdcTest.do_nothing_interp (3 ms)
[ RUN      ] XdcTest.unknown_function
[       OK ] XdcTest.unknown_function (3 ms)
[----------] 2 tests from XdcTest (7 ms total)
[----------] 3 tests from QuadTreeTest
[ RUN      ] QuadTreeTest.insert_bound_checking
[       OK ] QuadTreeTest.insert_bound_checking (0 ms)
[ RUN      ] QuadTreeTest.insert_count
[       OK ] QuadTreeTest.insert_count (1217 ms)
[ RUN      ] QuadTreeTest.insert_retrieve_same
[       OK ] QuadTreeTest.insert_retrieve_same (80 ms)
[----------] 3 tests from QuadTreeTest (1297 ms total)
[----------] Global test environment tear-down
[==========] 5 tests from 2 test cases ran. (1304 ms total)
[  PASSED  ] 5 tests.
```

**Pull request merged**